runtime/commonMain/src/kotlinx/serialization/SerialKinds.kt

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright 2017-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization
+
+/**
+ * Serial kind is an intrinsic property of [SerialDescriptor] that indicates how
+ * the corresponding type is structurally represented by its serializer.
+ *
+ * Kind is used by serialization formats to determine how exactly the given type
+ * should be serialized. For example, JSON format detects the kind of the value and,
+ * depending on that, may write it as a plain value for primitive kinds, open a
+ * curly brace '{' for class-like structures and square bracket '[' for list- and array- like structures.
+ *
+ * Kinds are used both in runtime, to serialize a value properly and statically,
+ * to retrospect the type structure or build serialization schema.
+ *
+ * Kind should match the structure of the serialized form, not the structure of the corresponding Kotlin class.
+ * Meaning that if serializable class `class IntPair(val left: Int, val right: Int)` is represented by the serializer
+ * as a single `Long` value, its descriptor should have [PrimitiveKind.LONG] without nested elements even though the class itself
+ * represents a structure with two primitive fields.
+ */
+public sealed class SerialKind {
+    override fun toString(): String {
+        // KNPE should never happen, because SerialKind is sealed and all inheritors are non-anonymous
+        return this::class.simpleName()!!
+    }
+}
+
+/**
+ * Values of primitive kinds that usually are represented as a single value.
+ * All default serializers for Kotlin [primitives types](https://kotlinlang.org/docs/tutorials/kotlin-for-py/primitive-data-types-and-their-limitations.html)
+ * and [String] have primitive kind.
+ *
+ * ### Serializers interaction
+ *
+ * Serialization formats typically handle these kinds by calling a corresponding primitive method on encoder or decoder.
+ * For example, if the following serializable class `class Color(val red: Byte, val green: Byte, val blue: Byte)` is represented by your serializer
+ * as a single [Int] value, a typical serializer will serialize its value in the following manner:
+ * ```
+ * val intValue = color.rgbToInt()
+ * encoder.encodeInt()
+ * ```
+ * and a corresponding [Decoder] counterpart.
+ *
+ * ### Implementation note
+ *
+ * Serial descriptors for primitive kinds are not expected to have any nested elements, thus its element count should be zero.
+ * If a class is represented as a primitive value, its corresponding serial name *should not* be equal to the corresponding primitive type name.
+ * For the `Color` example, represented as single [Int], its descriptor should have [INT] kind, zero elements and serial name **not equals**
+ * to `kotlin.Int`: `PrimitiveDescriptor("my.package.ColorAsInt", PrimitiveKind.INT)`
+ */
+public sealed class PrimitiveKind : SerialKind() {
```

Comment:
While writing the documentation, I've found a few problems related to our mental model: some kinds do not match each other.

So I propose the following breaking change:
* Make `UnionKind.OBJECT` structure kind instead, because it is not a union in any interpretation of "tagged union"
* Make `PolymorphicKind` a subclass of `UnionKind` because it is, in fact, a union. `SEALED` is just a more specific `ENUM` and `OPEN` is just a more specific `CONTEXTUAL`